src/foam/nanos/blob/FileService.java

```diff
@@ -0,0 +1,80 @@
+/**
+ * @license
+ * Copyright 2018 The FOAM Authors. All Rights Reserved.
+ * http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+package foam.nanos.blob;
+
+import foam.blob.Blob;
+import foam.blob.BlobService;
+import foam.blob.IdentifiedBlob;
+import foam.core.X;
+import foam.dao.DAO;
+import foam.nanos.fs.File;
+import org.apache.commons.io.IOUtils;
+
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import java.io.OutputStream;
+
+public class FileService
+    extends HttpBlobService
+{
+  protected DAO fileDAO_;
+  protected DAO userDAO_;
+
+  public FileService(X x, BlobService delegate) {
+    super(x, delegate);
+    fileDAO_ = (DAO) x.get("fileDAO");
+    // use the user dao instead of local user dao
+    // so that we get the authentication decoration
+    userDAO_ = (DAO) x.get("userDAO");
+  }
+
+  @Override
+  protected void download(X x) {
+    OutputStream os = null;
+    HttpServletRequest  req  = x.get(HttpServletRequest.class);
+    HttpServletResponse resp = x.get(HttpServletResponse.class);
+
+    try {
+      String path = req.getRequestURI();
+      String id = path.replaceFirst("/service/" + nspec_.getName() + "/", "");
+
+      // find file from file dao
+      File file = (File) fileDAO_.find_(x, id);
+      if ( file == null || !(file.getData() instanceof IdentifiedBlob) ) {
+        resp.setStatus(HttpServletResponse.SC_NOT_FOUND);
+        return;
+      }
+
+      // check to see if current user has access to file owner
+      if ( userDAO_.find_(x, file.getOwnerId()) == null ) {
+        resp.setStatus(HttpServletResponse.SC_UNAUTHORIZED);
+        return;
+      }
+
+      // get blob and blob size
+      String blobId = ((IdentifiedBlob) file.getData()).getId();
+      Blob blob = getDelegate().find_(x, blobId);
```

Comment:
I believe you can just call file.getData().read().

IdentifiedBlob has is a ProxyBlob with a factory on `delegate` that does the find internally.

You may have to call setX() on file.getData() because it might not be coming out of the DAO in the right context.  Although that's something we should consider fixing as well if its true.

Reply:
You're right. The Blob doesn't have it's context set

Reply:
The delegate factory doesn't seem to be working. I'm getting a null pointer exception when trying to get the size of the blob. And it's because the delegate is not set (I think)

Reply:
Interseting.  Leave a TODO here and I'll check it out later.